docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -569,80 +294,279 @@ helm template --values /tmp/values.yaml .
   {
     "name": "values-files",
-    "type": "enum",
-    "uiConfig": "{\"values\": [\"values.yaml\"]}"
+    "title": "Values Files",
+    "collectionType": "array"
   },
   {
-    "name": "image",
-    "type": "image",
-    "section": "Helm Parameters"
+    "name": "helm-parameters",
+    "collectionType": "map",
+    "title": "Helm Parameters",
+    "tooltip": "Parameters to override when generating manifests with Helm",
+    "map": {
+      "image.repository": "my.company.com/gcr-proxy/heptio-images/ks-guestbook-demo",
+      "image.tag": "0.1"
+    }
   }
 ]
 ```
 
+This is the full parameters announcement schema as Go types.
+
+```go
+package cmp
+
+// ParameterItemType is the primitive data type of each of the parameter's value (or each of its values, if it's an array or
+// a map).
+type ParameterItemType string
+
+const (
+	ParameterTypeString  ParameterItemType = "string"
+	ParameterTypeNumber  ParameterItemType = "number"
+	ParameterTypeBoolean ParameterItemType = "boolean"
+)
+
+// ParameterCollectionType is a parameter's value's type - a single value (like a string) or a collection (like an array or a
+// map).
+type ParameterCollectionType string
+
+const (
+	ParameterValueTypeString ParameterCollectionType = "string"
+	ParameterValueTypeMap    ParameterCollectionType = "map"
+	ParameterValueTypeArray  ParameterCollectionType = "array"
+)
+
+// ParameterAnnouncement represents a CMP's announcement of one acceptable parameter (though that parameter may contain
+// multiple elements, if the value holds an array or a map).
+type ParameterAnnouncement struct {
+	// Name is the name identifying a parameter. (required)
+	Name string                            `json:"name,omitempty"`
+	// Title is a human-readable text of the parameter name. (optional)
+	Title    string                        `json:"title,omitempty"`
+	// Tooltip is a human-readable description of the parameter. (optional)
+	Tooltip  string                        `json:"tooltip,omitempty"`
+	// Required defines if this given parameter is mandatory. (optional: default false)
+	Required bool                          `json:"required,omitempty"`
+	// ItemType determines the primitive data type represented by the parameter. Parameters are always encoded as
+	// strings, but ParameterTypes lets them be interpreted as other primitive types.
+	ItemType ParameterItemType             `json:"itemType,omitempty"`
+	// CollectionType is the type of value this parameter holds - either a single value (a string) or a collection (array or map).
+	// If Type is set, only the field with that type will be used. If Type is not set, `string` is the default. If Type
+	// is set to an invalid value, a validation error is thrown.
+	CollectionType ParameterCollectionType `json:"collectionType,omitempty"`
+	String         string                  `json:"string,omitempty"`
+	Map            map[string]string       `json:"map,omitempty"`
+	Array          []string                `json:"array,omitempty"`
```

Comment:
Maybe it is more intuitive to group those 3 attributes into a `DefaultValues` one?

Reply:
I think it should be reasonably obvious by looking at examples. I marginally prefer the flat style because it matches the Application manifest (i.e. is one less thing the CMP developer has to remember). But I could see a case for going either direction.